### ISSUE TITLE
[java] VariableNamingConventions doesn't check final variables as expected

### ISSUE THREAD

**Issue by eikemeier** [pmd/pmd#953]
**Affects PMD Version:** 6.1.0

**Rule:** [VariableNamingConventions](https://pmd.github.io/pmd-6.1.0/pmd_rules_java_codestyle.html#variablenamingconventions)

**Description:**

[VariableNamingConventionsRule.java](https://github.com/pmd/pmd/blob/master/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/codestyle/VariableNamingConventionsRule.java#L194) checks for `isUpperCase(varName.charAt(0))`, which is kind of expected, although [not documented](https://pmd.github.io/pmd-6.1.0/pmd_rules_java_codestyle.html#variablenamingconventions). This check is only done for non-final variables, giving unexpected results when combined with [MethodArgumentCouldBeFinal](https://pmd.github.io/pmd-6.1.0/pmd_rules_java_codestyle.html#methodargumentcouldbefinal) or [LocalVariableCouldBeFinal](https://pmd.github.io/pmd-6.1.0/pmd_rules_java_codestyle.html#localvariablecouldbefinal).

**Code Sample demonstrating the issue:**

```java
package …;

public final class Naming {
    private Naming() {
    }

    public static void function1(String Parameter) {
        String Local = Parameter;
    }

    public static void function2(final String Parameter) {
        final String Local = Parameter;
    }
}
```

While `Parameter` and `Local` are flagged in `function1`, they are not in `function2`:

```
Naming.java:7: Parameter 'Parameter' is not assigned and could be declared final
Naming.java:7: Variables should start with a lowercase character, 'Parameter' starts with uppercase character.
Naming.java:8: Local variable 'Local' could be declared final
Naming.java:8: Variables should start with a lowercase character, 'Local' starts with uppercase character.
```

**Running PMD through:** *Gradle*

**[Code Style](https://pmd.github.io/pmd-6.1.0/pmd_rules_java_codestyle.html)** enabled.

**Comment by oowekyala** [comment]
@eikemeier I don't really see unexpected behaviour here. Are you saying that `Parameter` and `Local` should be flagged in `function2`, or that they should not be flagged in `function1` since they could be made final?

**Comment by jsotuyod** [comment]
@oowekyala he points out that the final flag should be irrelevant. Arguments and local vars remain being arguments or local vars even if marked as final, so it makes no sense to not analyze them in the naming rule

https://github.com/pmd/pmd/blob/c4cf90e4cbc8af63056836f47640cea8fd77e3b2/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/codestyle/VariableNamingConventionsRule.java#L191-L204

**Comment by oowekyala** [comment]
Oh ok I understand now, thanks :) This shouldn't take long to fix

**Comment by jsotuyod** [comment]
The `VariableNamingConventions` rule is being deprecated in PMD 6.7.0 (#1303).

The new rule, `FormalParameterNamingConventions`, introduced in PMD 6.6.0 already handles the scenario here described correctly.

Therefore, I'm marking this as resolved. Users should migrate to the new rules as soon as possible.
